### ISSUE TITLE
Hw 4 Proposal 

### ISSUE THREAD

**Issue by ghost** [mikeizbicki/ucr-cs100#1486]
For the following proposal, I've have reconsidered my stance on it. To avoid 'reinventing the wheel,' I plan to document the gitlearn scripts for better references and better explanation on what each script's purpose of existence.
## Documentation
-  What is the purpose of this project:  Besides the config.sh being the most documented script on the repository, the rest of scripted are not well documented or are briefly summarized at the top of the page and would take time to actually sit down and read to understand.
-  Why would I do this?:  I wish to further document the repository so that a outside viewer (basic linux level) can follow what each line and parameter that the script runs.
-  How this documentation be used?:  I would document how each line proceeds upon being called by a user. Also, if someone else were to use gitlearn, they would be further understand more of the internal workings of each script.
## Webscript
-  Why make a webscript?: Because the current webscript for the repository looks like a semi-functional skeleton
-  How would the web script verify the user login?: use a Github PHP api to verify the login. 
-  Why do it this way?:  I'm trying to get it to the point where student's would only have a read access permission.  It would lower future temptations to try and change one's grade if they do not have a write permission available to them.
-  How would this benefit users?: It would benefit the user by having the information centralized on a single webpage than having to scout around the fork repository for the detailed version of the graded assignment.

**Comment by mikeizbicki** [comment]
You're proposing WAY too much.  I'd guess it would take me at least 200hrs to complete all that.
1. If you want to do this, then I would need to see some sort of proposal within the next week of how to handle the public/private keys.  There's a lot of details that need working out, and it will take some careful consideration.
2. A Qt-based application would not be helpful to anyone.  They're too much work to setup for nonprogrammers.  A fine project would be to document and extend the existing web-based solution we talked about earlier, but again I'd need more detail on specifically what you want to do.
3. I don't understand.  You can already download files from github.
4. How would your solution be better than the github issue tracker?  And there's already so much discussion board software there's no way you'd be able to implement something better.

**Comment by aso001** [comment]
Edit #1 above ... 
Decided to remove file/message board feature. 
Attempted to justify the rsa/gui proposal.

**Comment by mikeizbicki** [comment]
I'm having trouble reading your edits above.  Can you reformat using Markdown?  Also, in the future, you never want to make such major edits to a post like that.  The comments are (sadly) not in a git repo, so it makes it hard to follow along the conversation.

**Comment by aso001** [comment]
Edited again. I hope this is a readable format  this time around. Sorry for the inconvenience.

**Comment by mikeizbicki** [comment]
About RSA:
1. It looks like you want to implement your own version of the RSA protocol.  It's incredibly difficult to do this correctly.  Existing implementations have dozens of phds working on them and are still full of bugs.  Gitlearn currently uses the `gnupgp` program, so that's what you should use as well.
2. You have a lot of text about how to verify that a grade was submitted correctly.  But this is already done using public key signatures.  This is a feature built-in to git.  If you can change your grade right now without me noticing then you will get an unlimited amount of extra credit.
3. You haven't addressed any of the hard issues involving encryption.  In particular, where will the public and private keys be stored?  What would be the precise encryption protocol that will let me read and edit the grades, students read their own grades without editing, and no one else read?  Answering these questions is the hard part about this problem.
4.  Github does give all students free private repos: https://education.github.com/pack

GUI:
1. I don't want a desktop application GUI.  No one would use it because you have to install it and installing things is a pain.  In general, terminal applications and servers are the only things people are willing to install locally.  Everything else should be a webpage. I would accept something related to improving the webpage, but not a standalone application.

**Comment by ghost** [comment]
Redid the proposal again ...

**Comment by mikeizbicki** [comment]
Again, please don't edit your posts like this.  Always just reply with a new comment.  Editing should only be for minor spelling corrections.  It's very hard to follow a conversation when half of it gets deleted.

---

Documentation:
1.  GitLearn definitely needs more documentation, so I like this idea a lot.
2. The internals of the scripts are actually very well documented.  Adding more internal documentation would probably actually be bad.  The documentation would easily become out of date with the actual code.
3. The sort of documentation that GitLearn needs is better high-level documentation about how to run a course using gitlearn.  This documentation would take the form of a step-by-step tutorial that other instructors could use to setup a course that uses GitLearn.  A number of instructors have expressed interest in using GitLearn to manage their courses, but there's not enough documentation for them to be able to.  So this would have a direct real-world application.  Doing just this would be a project for at least 2 people (possibly more).

Website:
1. How is there temptation right now?  If you can change your grades right now without me knowing, I WANT you to do it!  That's a bug and bugs need to get fixed.
2. The current script already has login capabilities.
3. There is nothing in the GitLearn scripts that requires the use of github at all.  If I didn't want you to have write permissions to the repo, I could easily setup a git server on hammer that you could only read from and everything would work just fine.
4. If you're still interested in doing something with the website, we'll have to talk in person about it.

**Comment by ghost** [comment]
I will remember the next time. For now, I'm planning on documenting Gitlearn. I will talk to you about the website sometime before class or after class tomorrow.

**Comment by mikeizbicki** [comment]
Sounds great!

**Comment by ghost** [comment]
Second proposal(just in case):  
I would like to make grading program that grades output for labs and assignments. It would be similar to galah but usable for CS14 and higher. The steps it would take to process.  
1. Input and its corresponding output.    
-  It would pipe a test case to test if the program ran properly.  
-  The test case file would have follow a format: what it's testing for, the general responses, specific key words that its looking for, and a grade scale for each test case.  
  2.Analysis the output with a database.  
-  It would use the capture output do a diff check with the solutions previously imputed.  
-  Also, it would check with a previously submitted student submission and check for plagiarism  
- Evaluate the performance of the program.  
-  it could integrated the check for system calls  
- check for possible segmentation faults  
  4.Give a score of the performance of a program.  
-  Based on the number of differences  
-  Output a results texts based on the criteria based on the test cases  
-  What purpose would it serve to an open community: It would provide an automatic testing script for UNIX programming.  
-  Why wold it be a good idea: It would reduce human labor in grading assignments and allow me to practice utilize what I have learned in this class to do so. Such as piping and redirection along with databases from cs14.
-  How: It would be a computer operated with piping and redirection of input and output. It would compare the results of the output with the standard solution/results.

**Comment by mikeizbicki** [comment]
This is already being worked on.  See #1484 and #1496.

**Comment by mikeizbicki** [comment]
I also want to thank you for the detail and effort you put into describing each proposal you made.  That made it very easy for me to understand :)

**Comment by ghost** [comment]
So basically is this topic proposal locked to those to two groups?

**Comment by mikeizbicki** [comment]
No.  I think there'd be work for you on that topic too.  That'd be a lot of people, though; so I'd need a detailed breakdown of what exactly everyone plans to do.

**Comment by ghost** [comment]
### Objective:  An Automatic Computerized Grading System to be Implemented

## Working Tasks:

# 1. Input
-  The user would submit their source files for grading.  
-   When uploaded, the source files are in the testing folder, the code would be generated by using the users Makefile to compile. If errors/warnings were outputted onto a console, they would be captured as a data file and analyzed as factors for final grade evaluation.  
-  Before starting the test, an user's software program would be run with an internal check with checksyscalls.sh and cppchecker.sh to make sure that the required functions are already implemented in its code, and would be noted for their grading factors ... severity levels if the problem is found. With some pre-definited rules, an user's program  could be rejected whenever the applied occurrence found such as a single error found after the program compiled that a program fails to run correctly.  The rules could be a wrong return code value such as a zero, instead of a negative number.
- The testing pattern codes would be relied on pre-made cases which are made by the instructor/coder using a text editor to build. The patterns including a special string name such as "A Magic Word", or a special numeric value such "911" may be chosen for testing each program. Some patterns may be used to measure the time of execution and so on which are required for the student's program to accomplish.  For instance, the execution time could be chosen such 1 msec as a threshold to measure the performance of a software program written users.

# 2. Output
-  For each test case, an user's program would handle the pre-defined input test cases and generate the output captured as a data file when the program is running with each case. One or more data files may be generated, it depends on the user's program difficulty in terms of learning level.
- If the program results in a hidden segmentation fault or program bug, it would be terminated with the proposed project testing software program, and would be returned back to the user's email box as a respond, or the user may read his or her email to find out when the program is submitted.

# 3. Evaluation
-  A score file will be generated based on the outcome of data files during the phase of testing. Some supportive comments or hints may be shown to the user in order to encourage the user to learn more instead of giving up. The way to calculate the user's score may be based on the degree of matching numbers or patterns by comparing with the pre-defined data files made by an educator, or instructor. 

# 4. Score
-  A score file will be generated to keep track of all the achievement of the user group. It would be useful for an educator to show his teaching effort and students' academic level which could be used for the future students in the next quarter.  

## Testing Technique
-  Piping and Redirection are primarily used for communication and testing.  
-  diff() or strcmp() would be used as well to check the difference between the program output and the base testing case.  

## Possible additional features
- Automatic test script builder that asks an educator certain questions in a specific way on what he/she would like to test and place the compiled test case into a test case folder for an assignment/lab.  
-  A custom configuration setting could be easily implemented for an educator to optimize his/her specific requirements from being easy to strict on certain testing cases.  

## Note
-  I wish I could have an opportunity to work on this project I consider very important because it would be  great experience and be a nice addition on my resume when submitting to my future working company.

**Comment by mikeizbicki** [comment]
You can work on improving the existing testing program.  I've linked the issues above.  But I want all the effort focused on that one system and not have two separate systems.

**Comment by ghost** [comment]
Are you referring to gitlearn?

**Comment by mikeizbicki** [comment]
One of the components of gitlearn is an automated grading system.  A few different versions were created the past two quarters trying out different designs.  We've found a pretty good design that just needs some polishing up for it to be ready to deploy to a class.  We can talk in person about it on Friday if you want.

**Comment by ghost** [comment]
Ok I'll stay after on Friday to discuss more about it in detail.

**Comment by ghost** [comment]
Sorry. I was wondering if you were referring to gitlearn specifically or was there a specific repository related to grading aspects for assignments that specifically checks a student's assignment for errors. i ask because I don't know which issues link you were referring to?

**Comment by ghost** [comment]
We've setup a repo to hold all of the documentation, along with the dummy examples:
https://github.com/aso001/gitlearn_doc

Contributors: Nichelle(nchun003) and Angelo(aesgu001)

**Comment by mikeizbicki** [comment]
have you been able to run any of the scripts on it?

**Comment by ghost** [comment]
We are still working on it. We just ended our group call for night after trying to get most of the logistical planning done for who does what for everyone repos. We set everyone up to be a contributor for our repo/forks. However, we have yet to run any of the scripts atm. We plan to try each of the scripts after we finish setting up the 'dummy class' to test the script and then document the work by tmrw.
